Refactor block-inspector: improve maintainability and readability

getdave opened the pull request:
## What

This PR refactors the `block-inspector/index.js` component to eliminate duplication, improve code readability, and reduce maintenance burden while preserving all existing functionality.

**Key Changes:**

-   Extract duplicated `InspectorControls.Slot` patterns into `StyleInspectorSlots` component
-   Extract content panel logic into `ContentPanel` component
-   Inline both components to reduce file complexity
-   Eliminate `borderPanelLabel` prop drilling by calculating internally
-   Simplify conditional logic with well-named inline variables
-   Consolidate duplicate `useInspectorControlsTabs` calls for better performance
-   Improve variable names (`count` → `selectedBlockCount`)
-   Remove unnecessary API noise (`hasMultipleTabs` prop)

## Why

The `block-inspector/index.js` component had several maintainability issues that made it difficult to work with:

### **1. Significant Code Duplication**

-   **30+ lines of identical `InspectorControls.Slot` patterns** were duplicated in two places (multi-selection and single block rendering)
-   **Duplicate `useInspectorControlsTabs` calls** with identical parameters
-   **Similar conditional rendering patterns** repeated throughout the component

### **2. Poor Code Readability**

-   **Unclear variable names** like `count` (count of what?)
-   **Complex boolean expressions** that required mental parsing
-   **Inconsistent conditional logic** with similar patterns expressed differently

### **3. Performance Inefficiencies**

-   **Duplicate hook calls** - `useInspectorControlsTabs` called twice with same parameters
-   **Unnecessary prop drilling** - calculated values passed down instead of computed locally

### **4. Maintenance Burden**

-   **Changes required in multiple places** - style slot modifications needed updates in two locations
-   **Tight coupling** - components dependent on parent calculations
-   **API complexity** - unnecessary props being passed around

## How

### **Eliminated Duplication**

-   Created `StyleInspectorSlots` component to consolidate all style-related inspector control slots
-   Created `ContentPanel` component to encapsulate content panel logic
-   Inlined both components to reduce file complexity while maintaining reusability

### **Improved Readability**

-   Replaced complex boolean expressions with well-named variables:
    -   `count > 1 && ! isSectionBlock` → `isMultiSelection`
    -   `! blockType || ! selectedBlockClientId || isSelectedBlockUnregistered` → `shouldShowWarning`
-   Renamed unclear variables: `count` → `selectedBlockCount`

### **Optimized Performance**

-   Consolidated `useInspectorControlsTabs` calls - now called once in parent and passed down
-   Eliminated unnecessary prop drilling by having components calculate their own derived values

### **Simplified APIs**

-   Removed `hasMultipleTabs` prop - child component calculates this locally
-   Eliminated `borderPanelLabel` prop drilling - component calculates its own label

## Testing

-   All existing functionality preserved
-   No visual or behavioral changes
-   All conditional rendering logic works identically
-   Performance improved with fewer hook calls

To confirm existing functionality is preserved ensure the displayed tabs and controls for this branch match trunk using:
- a section block, template parts, unsynced patterns, and synced patterns
- a regular individual block selection
- a mutli-selection with only non-section blocks
- a multi-selection including a section block

Additionally, 

## Rationale for Value

This refactoring provides **genuine value** by solving real problems developers face:

1. **Maintenance Efficiency** - Style slot changes now only need to be made in one place
2. **Developer Experience** - Code reads like plain English with self-documenting variable names
3. **Runtime Performance** - Eliminated duplicate hook calls
4. **Reduced Coupling** - Components are more self-contained and easier to reason about
5. **Cleaner Interfaces** - Only essential data is passed between components
6. **Future-Proofing** - Makes the component more open to change for upcoming features

### **Preparing for Simplified Editing Experience (WP 6.9)**

This refactoring is particularly important as we prepare for the **Simplified Editing Experience** work planned for WordPress 6.9. The current monolithic structure of the block inspector would make it difficult to:

-   **Modify inspector behavior** for different editing modes
-   **Add new conditional rendering logic** for simplified interfaces
-   **Extract or reorganize inspector panels** based on user preferences
-   **Implement dynamic inspector layouts** that adapt to the editing context

By breaking the component into smaller, focused pieces with clear responsibilities, we create a foundation that can more easily accommodate the architectural changes needed for the simplified editing experience.

This is **not refactoring for its own sake** - each change addresses a specific maintainability or performance issue that developers encounter when working with this component, while also preparing the codebase for future enhancements.

## Screenshots

No visual changes - this is a pure refactoring that preserves all existing functionality.

getdave committed:
Consolidate the Styles slots

getdave committed:
Refactor block-inspector: extract and inline StyleInspectorSlots and ContentPanel components

- Extract duplicated InspectorControls.Slot patterns into StyleInspectorSlots component
- Extract content panel logic into ContentPanel component
- Inline both components into main index.js file to reduce file complexity
- Eliminate borderPanelLabel prop drilling by calculating it internally
- Remove separate component files (style-inspector-slots.js, content-panel.js)
- Improve code cohesion and maintainability

getdave committed:
Consolidate vars

getdave committed:
Remove unused onSelect prop from ContentPanel

The onSelect prop was never passed to ContentPanel and was always undefined.
This cleans up the API by removing the unused parameter.

getdave commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -39,9 +39,71 @@ function BlockStylesPanel( { clientId } ) {
+			/>
+		</PanelBody>
+	);
+}
```
Done

getdave commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -242,67 +278,36 @@ const BlockInspectorSingleBlock = ( {
-							/>
-						</PanelBody>
-					) }
+					<ContentPanel contentClientIds={ contentClientIds } />
```
Done

jeryj commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -20,11 +20,11 @@ import BlockVariationTransforms from '../block-variation-transforms';
-import AdvancedControls from '../inspector-controls-tabs/advanced-controls-panel';
+import { default as InspectorControls } from '../inspector-controls';
 import PositionControls from '../inspector-controls-tabs/position-controls-panel';
+import AdvancedControls from '../inspector-controls-tabs/advanced-controls-panel';
```
Nit: I think these lines just got moved and not changed.

aaronrobertshaw replied:
Done in https://github.com/WordPress/gutenberg/pull/71608/commits/0b18c3b0030194963f89cb99448741dae6163128

jeryj commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -64,7 +123,7 @@ function BlockInspector() {
 
 		return {
-			count: getSelectedBlockCount(),
+			selectedBlockCount: getSelectedBlockCount(),
```
I think this is a good rename. Makes it clearer what we're counting.

jeryj commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -73,7 +132,7 @@ function BlockInspector() {
 
 	const availableTabs = useInspectorControlsTabs( blockType?.name );
-	const showTabs = availableTabs?.length > 1;
+	const hasMultipleTabs = availableTabs?.length > 1;
```
I like `showTabs` or `showTabbedInterface` more than `hasMultipleTabs`. While `hasMultipleTabs` is true, I think the root condition we need to know here is if we use the tabbed interface or dump all the panels directly in the sidebar without the tabbed interface.

jeryj commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -84,42 +143,16 @@ function BlockInspector() {
-	const borderPanelLabel = useBorderPanelLabel( {
-		blockName: selectedBlockName,
-	} );
+	const isMultiSelection = selectedBlockCount > 1 && ! isSectionBlock;
```
Why the `! isSectionBlock` here?

aaronrobertshaw replied:
To match the original check below:

`count > 1 && ! isSectionBlock`

jeryj commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -132,11 +165,10 @@ function BlockInspector() {
-		! selectedBlockClientId ||
-		isSelectedBlockUnregistered
-	) {
+	const shouldShowWarning =
+		! blockType || ! selectedBlockClientId || isSelectedBlockUnregistered;
+
+	if ( shouldShowWarning ) {
```
I'm fine moving this to a named `const` to define what we're checking, but `shouldShowWarning` doesn't feel appropriate when the message is "No block selected." Maybe, `noBlockSelection`? But, oddly, `isSelectedBlockUnregistered` could show a "No block selected" message?

aaronrobertshaw replied:
Neither seem to be a great fit at the moment so perhaps we can leave as is for now?

jeryj commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -39,9 +39,68 @@ function BlockStylesPanel( { clientId } ) {
+	showPositionControls = true,
+	showListControls = false,
+	showBindingsControls = true,
+	className,
```
`className` only gets used on the color slot, so it doesn't feel like a well-named prop. I wonder if it's even necessary.

aaronrobertshaw replied:
I think this can be removed, especially as it isn't reused across slots and doesn't make a lot of sense.

aaronrobertshaw replied:
Done via https://github.com/WordPress/gutenberg/pull/71608/commits/9784738b8d62b9fdccf1338621a7110a9d25c014

jeryj commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -39,9 +39,68 @@ function BlockStylesPanel( { clientId } ) {
 
+function StyleInspectorSlots( {
+	blockName,
+	showAdvancedControls = true,
+	showPositionControls = true,
+	showListControls = false,
+	showBindingsControls = true,
```
I think these can be removed. If a slot is empty, I think it may not render at all. If it does, maybe a fillProps.length check would be a better fix.

aaronrobertshaw replied:
This prop will need to be kept to fix a bug introduced by this refactor.

After the `StyleInspectorSlots` component was introduced and used for the multi-selection mode, it now incorrectly renders advanced, position, and bindings controls.

jeryj commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -207,9 +240,10 @@ const BlockInspectorSingleBlock = ( {
 } ) => {
-	const availableTabs = useInspectorControlsTabs( blockName );
-	const showTabs = ! isSectionBlock && availableTabs?.length > 1;
+	const hasMultipleTabs = availableTabs?.length > 1;
+	const shouldShowTabs = ! isSectionBlock && hasMultipleTabs;
```
We could use this same named logic for the above `showTabs`. Having naming consistency would be helpful.

aaronrobertshaw replied:
Personally, I don't the naming changes make much of a difference either way. As it stands in this PR, the `hasMultipleTabs` name is pretty literal, the changes here are consistent naming and check wise against the earlier `hasMultipleTabs`.

I'm inclined to leave this as is.

jeryj commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -39,9 +39,68 @@ function BlockStylesPanel( { clientId } ) {
+			<BlockQuickNavigation clientIds={ contentClientIds } />
+		</PanelBody>
+	);
+}
```
I think we can undo this becoming a function. It's only used one time at the moment, and I think it's easier to read as the existing if statement rather than showing it as rendered and then having to dig into this function to see if it actually gets output or not.

aaronrobertshaw replied:
Done via https://github.com/WordPress/gutenberg/pull/71608/commits/f5e5aaec824226c20cdcdb1549705d1648c7fb35

jeryj commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -162,6 +194,7 @@ function BlockInspector() {
 				clientId={ selectedBlockClientId }
 				blockName={ blockType.name }
 				isSectionBlock={ isSectionBlock }
+				availableTabs={ availableTabs }
```
I think this is a good refactor. We already hit the `useInspectorControlTabs` above, so no need to do it again within the `BlockInspectorSingleBlock`.

jeryj reviewed (commented):
I think it needs cleanup, but, overall, I like combining the Slots into one component rather than having the same thing unnecessarily duplicated.

aaronrobertshaw committed:
Fix multi-selection mode inspector slot rendering

aaronrobertshaw commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -84,42 +143,16 @@ function BlockInspector() {
-						/>
-						<InspectorControls.Slot group="styles" />
-					</>
+					<StyleInspectorSlots blockName={ selectedBlockName } />
```
This is missing the props for `showAdvancedControls`, `showPositionControls`, and `showBindingsControls`, so it now incorrectly renders these slots for multi-selections when it shouldn't.

aaronrobertshaw reviewed (commented):
Thanks for all the work in refactoring these slots @getdave and @jeryj 👍 

It might help to get additional eyes on this PR as it impacts several use cases. The test also instructions need to help reviewers as well (I'll update them soon).

To ensure existing functionality I compared this branch against trunk for:
- a section block, template parts, unsynced patterns, and synced patterns
- a regular individual block selection
- a mutli-selection with only non-section blocks
- a multi-selection including a section block

From testing it looks like there was a bug introduced for mutli-selections but it's an easy fix.

Overall, I think the changes are heading in the right direction. So to help move this along and progress https://github.com/WordPress/gutenberg/issues/71626, I'll push a few tweaks addressing current feedback.

aaronrobertshaw committed:
Remove ContentPanel component and inline its functionality

The ContentPanel component was only used in one place, so inline its
simple logic directly to reduce component complexity.

aaronrobertshaw committed:
Remove className prop from StyleInspectorSlots

aaronrobertshaw committed:
Restore import order

aaronrobertshaw commented:
I've pushed a fix and addressed the current feedback on this PR. I think it is ready for final review and test before merging.

I'll also be putting up a PR based on this one to introduce content tabs for https://github.com/WordPress/gutenberg/issues/71626 shortly.

andrewserong reviewed (approved):
> Thanks for all the work in refactoring these slots @getdave and @jeryj 👍

+1, and thanks for the updates @aaronrobertshaw, overall I like the idea of this refactor, too 👍

> To ensure existing functionality I compared this branch against trunk for:
> a section block, template parts, unsynced patterns, and synced patterns
> a regular individual block selection
> a mutli-selection with only non-section blocks
> a multi-selection including a section block

Cheers for this, these are all testing well for me, along with a block set up with block bindings, too (shows in single selection, not when part of a multiselection).

LGTM!

tellthemachines commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -92,50 +136,31 @@ function BlockInspector() {
+	if (
+		hasSelectedBlocks &&
+		! isSectionBlockInSelection &&
+		! isSectionBlock
```
Isn't it redundant to check whether `! isSectionBlock` as that would already be answered by `! isSectionBlockInSelection`?

andrewserong replied:
It's probably a bit edge-casey, but `isSectionBlock` can also refer to a parent of the selected block, so I think that might be why it's included in this check, whereas `isSectionBlockInSelection` is only about the selected blocks?

tellthemachines replied:
Hmmm, if I'm reading [`isSectionBlock`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/store/private-selectors.js#L512) correctly, it only checks the block it gets passed. It does look at the section root but only to determine whether that block is a direct child of it (which would mean it's a section). 

I guess ultimately if this addition _does_ change the behaviour, it doesn't belong in a PR that claims to be a straightforward refactor. And if it doesn't, there's no reason to add it 😄

andrewserong replied:
> Hmmm, if I'm reading [isSectionBlock](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/store/private-selectors.js#L512) correctly, it only checks the block it gets passed. It does look at the section root but only to determine whether that block is a direct child of it (which would mean it's a section).

In this case it's a little confusing, I was referring to the `isSectionBlock` const, not the function, which has the same name 😄. Above, `isSectionBlock` is defined by this call:

```
isSectionBlock: _isSectionBlock( renderedBlockClientId ),
```

And it's `renderedBlockClientId` that can be the parent:

https://github.com/WordPress/gutenberg/blob/6e368dd2322d706d020e514219eaf64b1eda3e8b/packages/block-editor/src/components/block-inspector/index.js#L105-L107

tellthemachines replied:
Oh I see.

Ok so the condition is:

* `hasSelectedBlocks` is true: there's more than one block selected.
* `! isSectionBlockInSelection`: hopefully this one speaks for itself 😅 
* `! isSectionBlock`: if `hasSelectedBlocks` is true, this is always going to be false, because `renderedBlockClientId` is null if there's more than one block selected. 

I think we should remove it for the sake of readability, particularly given this is a refactor meant to improve readability. I'm guessing it's likely to be an AI artifact rather than a human change 😅

aaronrobertshaw replied:
I'm a bit late to the discussion but if I'm following it correctly, the question is what's the difference between `isSectionBlockInSelection` and `isSectionBlock`?

I think @andrewserong touches on it, that the `renderedBlockClientId` could be the parent section rather than the currently selected block. So this `renderedBlockClientId` could be different than those in the set of currently selected blocks represented by `isSectionBlockInSelection`.

> I'm guessing it's likely to be an AI artifact rather than a human change

Probably more that I messed up the conflict resolution with https://github.com/WordPress/gutenberg/pull/71708 and the PR I had based on this (https://github.com/WordPress/gutenberg/pull/71714).

andrewserong replied:
> ! isSectionBlock: if hasSelectedBlocks is true, this is always going to be false, because renderedBlockClientId is null if there's more than one block selected.

Good catch! Yes it sounds like the `! isSectionBlock` in this and the below check are redundant because `isSectionBlock` will always be falsy if `hasSelectedBlocks` is true, like you mention 👍

aaronrobertshaw replied:
Thanks for the discussion here. I see the issue now, I'll clean this up shortly.

andrewserong replied:
Thanks @aaronrobertshaw! IMO once that's tidied up, this looks in good shape to merge to me.

aaronrobertshaw replied:
Done in [858e157](https://github.com/WordPress/gutenberg/pull/71608/commits/858e157df1ee2cf6c8b3c133bcb84a2c4064aa4e).

tellthemachines commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -92,50 +136,31 @@ function BlockInspector() {
 	}
 
-	if ( hasSelectedBlocks && isSectionBlockInSelection ) {
+	if ( hasSelectedBlocks && isSectionBlockInSelection && ! isSectionBlock ) {
```
Here `isSectionBlockInSelection` seems to contradict `! isSectionBlock`. Which one should it be? 😅

aaronrobertshaw replied:
~Both, if my understanding of the nuance around the `isSectionBlock` value potentially also including the parent section block is correct.~ Nevermind.

tellthemachines commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -71,7 +115,7 @@ function BlockInspector() {
 		);
```
Not a change from this PR but might be worth improving: the above `renderedBlockClientId` uses both `_selectedBlockClientId` and calls `getSelectedBlockClientId()` again. We could avoid calling the selector twice by just using `_selectedBlockClientId` instead.

aaronrobertshaw replied:
Done in https://github.com/WordPress/gutenberg/pull/71608/commits/24de8b3cb0c5f893f561290bc83b68be04e6a399

tellthemachines replied:
Nice, thanks!

tellthemachines reviewed (commented):
I can't find any changes in behaviour between this and trunk, but was curious about some of the logic so left questions below. Not blockers for merging though!

aaronrobertshaw committed:
Clean up rendundant checks

aaronrobertshaw committed:
Remove unnecessary selector call

tellthemachines reviewed (approved):
LGTM!

getdave commented:
Thanks for taking this over and merging (FYI I was AFK sick) 🙇
